### PR TITLE
Make Spark SQL APIs supported across all types

### LINKED ISSUE #1580: Eliminate unused ConcreteADAMRDDFunctions class
I was considering doing this in #1391, as it would eliminate the classtag hacking I've added in https://github.com/bigdatagenomics/adam/pull/1391/commits/92e296eb98c3ec585a4c365c37dd3f07827e9849. However, I don't want this to come in to #1391, because I'd rather it get a standalone review, etc. Here's the to-do:

- [ ] Eliminate unused ConcreteADAMRDDFunctions class
- [ ] Move functionality from ADAMRDDFunctions into GenomicRDD
- [ ] Change GenomicRDD from a trait into an abstract class with a view bound on `T: Manifest`, which eliminates the need to hack the classtags

### PR DESCRIPTION
**Pull request by fnothaft**
Resolves #1580, #1867, WIP towards #1728. Merges the GenomicRDD and GenomicDataset traits together, and makes the VariantContextRDD and GenericGenomicRDD implementations support GenomicDataset. Not done yet, but I wanted to get this out for review, as it's a big one. Remaining TODO:

- [x] Clear merge conflicts
- [x] Plumb join and GenericGenomicRDD APIs out to Python
- [x] Plumb join and GenericGenomicRDD APIs out to R
- [x] Resolve a serialization issue that occurs when using GenericGenomicRDD from the dataset API
- [x] Update user docs

One general stylistic question: this PR goes halfway to renaming everything "GenomicDataset" instead of "GenomicRDD". Is this a good change, or should we hold on to the GenomicRDD name for the sake of consistency?

### CODE REVIEW

**Comment by heuermh** [comment]
> One general stylistic question: this PR goes halfway to renaming everything "GenomicDataset" instead of "GenomicRDD".

+1

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMSaveAnyArgs.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * Argument configuration for saving any output format.
+ */
+trait ADAMSaveAnyArgs extends SaveArgs {
+
+  /**
+   * If true and saving as FASTQ, we will sort by read name.
+   */
+  var sortFastqOutput: Boolean
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMSaveAnyArgs.scala]: is now a good time to rid the world of this?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 2]
```diff
@@ -2246,6 +2287,49 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
     }
   }
 
+  /**
+   * Load a path name in Parquet + Avro format into a VariantContextRDD.
+   *
+   * @param pathName The path name to load genotypes from.
+   *   Globs/directories are supported.
+   * @return Returns a GenotypeRDD.
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: doc doesn't match

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 3]
```diff
@@ -2246,6 +2287,49 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
...
+    val sd = loadAvroSequenceDictionary(pathName)
+
+    // load avro record group dictionary and convert to samples
+    val samples = loadAvroSamples(pathName)
+
+    val sqlContext = SQLContext.getOrCreate(sc)
+    import sqlContext.implicits._
+    val ds = sqlContext.read.parquet(pathName).as[VariantContextProduct]
+
+    new DatasetBoundVariantContextRDD(ds, sd, samples, headers)
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: we're intentionally bound to dataset here, is the idea to move other loadParquetXxx methods in the same direction?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala [hunk 4]
```diff
@@ -111,13 +118,212 @@ private[rdd] object GenomicRDD {
...
- * @tparam U The type of this GenomicRDD.
+ * @tparam U The type of this GenomicDataset.
  */
-trait GenomicRDD[T, U <: GenomicRDD[T, U]] extends Logging {
+trait GenomicDataset[T, U <: Product, V <: GenomicDataset[T, U, V]] extends Logging {
+
+  val uTag: TypeTag[U]
+
+  /**
+   * This data as a Spark SQL Dataset.
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala]: These data?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala [hunk 5]
```diff
@@ -111,13 +118,212 @@ private[rdd] object GenomicRDD {
...
+   */
+  def transformDataset(tFn: Dataset[U] => Dataset[U]): V
+
+  /**
+   * Applies a function that transforms the underlying DataFrame into a new DataFrame
+   * using the Spark SQL API.
+   *
+   * @param tFn A function that transforms the underlying RDD as a DataFrame.
+   * @return A new RDD where the RDD of genomic data has been replaced, but the
+   *   metadata (sequence dictionary, and etc) is copied without modification.
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala]: metadata ... are?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala [hunk 6]
```diff
@@ -111,13 +118,212 @@ private[rdd] object GenomicRDD {
...
+   */
+  def transformDataFrame(tFn: DataFrame => DataFrame)(
+    implicit uTag: TypeTag[U]): V = {
+    val sqlContext = SQLContext.getOrCreate(rdd.context)
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala]: this method/class is deprecated in Spark 2.x, no?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala [hunk 7]
```diff
@@ -111,13 +118,212 @@ private[rdd] object GenomicRDD {
...
+   *   metadata (sequence dictionary, and etc) is copied without modification.
+   */
+  def transformDataFrame(tFn: JFunction[DataFrame, DataFrame]): V = {
+    val sqlContext = SQLContext.getOrCreate(rdd.context)
+    import sqlContext.implicits._
+    transformDataFrame(tFn.call(_))(uTag)
+  }
+
+  /**
+   * Applies a function that transmutes the underlying RDD into a new RDD of a
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala]: RDD &rarr; Dataset

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala [hunk 8]
```diff
@@ -604,24 +825,26 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] extends Logging {
...
-    pipe(cmd,
-      files.asInstanceOf[Seq[String]].toList,
+  def pipe[X, Y <: Product, Z <: GenomicDataset[X, Y, Z], W <: InFormatter[T, U, V, W]](cmd: String,
+                                                                                        files: Seq[Any],
+                                                                                        environment: java.util.Map[Any, Any],
+                                                                                        flankSize: java.lang.Double,
+                                                                                        tFormatter: Class[W],
+                                                                                        xFormatter: OutFormatter[X],
+                                                                                        convFn: Function2[V, RDD[X], Z]): Z = {
+    val jList: java.util.List[String] = files.asInstanceOf[Seq[String]].toList
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala]: are these `val`s actually necessary?  it read better before

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala [hunk 9]
```diff
@@ -924,55 +1134,51 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] extends Logging {
...
+   *   right RDD that did not overlap a key in the left RDD.
    *
-   * @see broadcastRegionJoinAgainstAndGroupByRight
+   * @see rightOuterBroadcastRegionJoin
    */
-  def broadcastRegionJoinAndGroupByRight[X, Y <: GenomicRDD[X, Y]](
-    genomicRdd: GenomicRDD[X, Y],
-    flankSize: Long)(
+  def rightOuterBroadcastRegionJoin[X, Y <: Product, Z <: GenomicDataset[X, Y, Z]](
+    genomicRdd: GenomicDataset[X, Y, Z])(
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala]: genomicRdd &rarr; genomicDataset

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala [hunk 10]
```diff
@@ -1018,16 +1240,49 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] extends Logging {
...
+        (kv._1.flatMap(broadcast.backingDataset.getReferenceRegions) ++
+          getReferenceRegions(kv._2))
+          .toSeq
+      },
+      kv => (kv._1.map(broadcast.backingDataset.productFn).toSeq, productFn(kv._2)),
+      kv => (kv._1.map(broadcast.backingDataset.unproductFn), unproductFn(kv._2)))
+  }
+
+  /**
+   * Performs a broadcast inner join between this RDD and another RDD.
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala]: RDD &rarr; Dataset

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala [hunk 11]
```diff
@@ -2030,23 +2331,23 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] extends Logging {
...
+  val regionFn: T => Seq[ReferenceRegion]
 
-  def union(rdds: GenericGenomicRDD[T]*): GenericGenomicRDD[T] = {
-    val iterableRdds = rdds.toSeq
-    GenericGenomicRDD(rdd.context.union(rdd, iterableRdds.map(_.rdd): _*),
-      iterableRdds.map(_.sequences).fold(sequences)(_ ++ _),
-      regionFn)
+  @transient val uTag: TypeTag[U] = typeTag[U]
+
+  def saveAsParquet(filePath: String,
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala]: does this one need an override?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala [hunk 12]
```diff
@@ -2377,12 +2657,61 @@ abstract class MultisampleAvroGenomicRDD[T <% IndexedRecord: Manifest, U <: Prod
...
+   *   true, delta encoding is used. Defaults to false.
+   * @param optSchema The optional schema to set. Defaults to None.
+   */
+  protected def saveRddAsParquet(
+    pathName: String,
+    blockSize: Int = 128 * 1024 * 1024,
+    pageSize: Int = 1 * 1024 * 1024,
+    compressCodec: CompressionCodecName = CompressionCodecName.GZIP,
+    disableDictionaryEncoding: Boolean = false,
+    optSchema: Option[Schema] = None): Unit = SaveAsADAM.time {
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala]: would be nice to have separate SaveRddAsParquet and SaveDatasetAsParquet timers

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/variant/GenotypeRDD.scala [hunk 13]
```diff
@@ -210,7 +210,10 @@ case class RDDBoundGenotypeRDD private[rdd] (
   }
 }
 
-sealed abstract class GenotypeRDD extends MultisampleAvroGenomicRDD[Genotype, GenotypeProduct, GenotypeRDD] {
+sealed abstract class GenotypeRDD extends MultisampleAvroGenomicDataset[Genotype, GenotypeProduct, GenotypeRDD] {
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/variant/GenotypeRDD.scala]: will you be proposing to move this to `GenotypeDataset`?

**Comment by fnothaft** [comment]
Jenkins, retest this please.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMSaveAnyArgs.scala (hunk 1 above)
Comment (fnothaft) [reply adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMSaveAnyArgs.scala]: Don't think so.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala (hunk 9 above)
Comment (fnothaft) [reply adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala]: How much do you want me to go through and patch these? For now, I've left them as is (genomicRdd). I'd rather make a cleanup pass through later.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala (hunk 12 above)
Comment (fnothaft) [reply adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala]: We're not really going to get meaningful timing info out of the `SaveDatasetAsParquet` timer.

**Comment by fnothaft** [comment]
@heuermh I think everything in Scala is going to be stable from here on out; can you make a review pass? I'm going to finish up the R bit in the morning.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala (hunk 3 above)
Comment (fnothaft) [reply adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: Nah, we don't have an Avro impl of VariantContext, so we can't use the normal loadParquet method.

Comment (heuermh) [reply adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: ok

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMSaveAnyArgs.scala (hunk 1 above)
Comment (heuermh) [reply adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMSaveAnyArgs.scala]: ok

File: scripts/jenkins-test [hunk 14]
```diff
@@ -213,9 +213,11 @@ then
 	    export SPARKR_SUBMIT_ARGS="--jars ${ASSEMBLY_DIR}/${ASSEMBLY_JAR} --driver-class-path ${ASSEMBLY_DIR}/${ASSEMBLY_JAR} sparkr-shell"
 	    
             # we can run the python build, now that we have a spark executable
+	    # -DskipTests appears to skip the python/r tests?
 	    mvn -U \
 		-P python,r \
 		test \
+		-Dsuites=select.no.suites\* \
```
Comment (heuermh) [inline scripts/jenkins-test]: is this to ignore the scala tests while still running the Python and R ones?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala (hunk 12 above)
Comment (heuermh) [reply adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala]: ok

**Comment by antonkulaga** [comment]
>Is this a good change, or should we hold on to the GenomicRDD name for the sake of consistency?

Do not know, for my ADAM-based code/libs renaming is not a problem.. What I noticed with the previous version was that in many cases using genomic RDDs was faster than trying to use spark sql.

**Comment by heuermh** [comment]
Thank you, @fnothaft

**Pull request merged**